packages/react-reconciler/src/ReactFiberCompleteWork.js

```diff
@@ -865,8 +866,36 @@ function completeWork(
       if ((workInProgress.effectTag & DidCapture) !== NoEffect) {
         // Something suspended. Re-render with the fallback children.
         workInProgress.expirationTime = renderExpirationTime;
+
         // Do not reset the effect list.
         return workInProgress;
+      } else if ((workInProgress.effectTag & SoftCapture) !== NoEffect) {
+        // something called avoidThisRender()
+        if (
+          current === null ||
+          (current.memoizedState &&
+            renderExpirationTime > current.memoizedState.didAvoidRenderTime)
+        ) {
+          // Something soft suspended, and this is the first "trigger rerender" pass.
+          // Re-render with the fallback children.
+          if (!workInProgress.memoizedState) {
+            // fallback isn't showing yet, trigger fallback pass
+            workInProgress.expirationTime = renderExpirationTime;
+            return workInProgress;
+          } else {
+            // TODO: we're never entering this branch because
+            // effectTag: SoftCapture is cleared after the previous pass
+            // does it matter?
```

Comment:
You could make it an invariant that workInProgress.memoizedState is null. If it's not, then there's a bug somewhere since any throwing/avoidThisRender in the fallback should bubble past this. That way you also don't have to add a fake `return null`.